File: src/update.ts
import {
  MoveObject,
  Players,
  Data,
  Board,
  GameRecord,
  Side,
} from "./interfaces";

export function updatePlayers(players: Players, currentPlayer: string) {
  if (!players[currentPlayer]) players[currentPlayer] = 1;
  else ++players[currentPlayer];
  return players;
}

export function updateMostRecentMoves(
  moveArray: MoveObject[],
  moveObj: MoveObject
) {
  moveArray.unshift(moveObj);
  return moveArray.slice(0, 3);
}

export function updateAfterTurn(data: Data, move: MoveObject) {
  let newData = { ...data };
  newData.players = updatePlayers(newData.players, move.name);
  ++newData.totalMoves;
  newData.mostRecentMoves = updateMostRecentMoves(
    newData.mostRecentMoves,
    move
  );
  return newData;
}

export function updateMostRecentGames(
  mostRecentGames: GameRecord[],
  board: Board
) {
  const gameRecord: GameRecord = {
    scores: {
      top: board.scores.top,
      bot: board.scores.bot,
    },
    turnsPlayed: board.turnsPlayed,
  };

  mostRecentGames.unshift(gameRecord);
  return mostRecentGames;
}

export function updateAfterGame(data: Data, board: Board) {
  const winner = board.currentTurn;
  ++data.totalGames;
  ++data.wins[winner];
  data.mostRecentGames = updateMostRecentGames(data.mostRecentGames, board);
  return data;
}

File: cmd/sungka/update.go
package main

func updatePlayers(players Players, currentPlayer string) Players {
	_, ok := players[currentPlayer]
	if !ok {
		players[currentPlayer] = 1
	} else {
		players[currentPlayer]++
	}
	return players
}

func updateMostRecentMoves(moveArray []MoveObject, moveObj MoveObject) []MoveObject {
	return append([]MoveObject{moveObj}, moveArray...)[0:3]
}

func updateAfterTurn(data Data, move MoveObject) Data {
	newData := data
	newData.players = updatePlayers(newData.players, move.name)
	newData.totalMoves++
	newData.mostRecentMoves = updateMostRecentMoves(newData.mostRecentMoves, move)
	return newData
}

func updateMostRecentGames(mostRecentGames []GameRecord, board Board) []GameRecord {
	gameRecord := GameRecord{
		scores: Scores{
			top: board.scores.top,
			bot: board.scores.bot,
		},
		turnsPlayed: board.turnsPlayed,
	}

	return append([]GameRecord{gameRecord}, mostRecentGames...)
}

func updateAfterGame(data Data, board Board) Data {
	winner := board.currentTurn
	data.totalGames++
	data.wins[winner]++
	data.mostRecentGames = updateMostRecentGames(data.mostRecentGames, board)
	return data
}
